### DISCUSSION TITLE
DDOP sending ?

### DISCUSSION THREAD

**Discussion by franz-ms-muc** [Open-Agriculture/AgIsoStack-plus-plus#300] [Ideas]
Hi, 

how do you plan to send the DDOP ? 
does it need to be generated externally with a seperate Tool ? 
if yes, do you know such Tools ? 
i only found this one: https://anedo.eu/produkt/ddopmaker/
do you know alternative ones ? 

or is the ddop generated dynamically ? 

Thanks, 
Franz

**Comment by ad3154** [comment]
Hello!

We support both generating it dynamically or supplying one created with such a tool.

One nice thing about generating it dynamically by using our `DeviceDescriptorObjectPool` class is that you can alter it at runtime to meet your needs, such as selecting different object descriptors depending on the language being used.

You can see an example of generating one programmatically in [this example](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/blob/91fd4915108fb8ccd5e841d753e19231f9e620d2/examples/task_controller_client/section_control_implement_sim.cpp#L71C40-L71C51) otherwise you can supply a binary DDOP to the TC client with `TaskControllerClient::configure`.

I actually think it would be very easy to create such a tool myself - it might take an afternoon or so. If I get the time I will see about doing that.

**Reply by ad3154** [reply]
I did start making a tool [located here](https://github.com/Open-Agriculture/AgIsoDDOPGenerator) but as I mentioned, it's not required. I do hope though that this will be a free alternative to such paid products 🙂

Either way, once you have a DDOP either created programmatically or with the tool, as long as you call `TaskControllerClient::configure` and provide the DDOP to the stack, it will automatically do the uploading for you when the partner you provide comes online.
